<a href="https://colab.research.google.com/github/gulamkibria775/Data_science_project/blob/main/Data_science_project/single_layer_percepton_gradient_optimizer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [21]:
import tensorflow as tf

tf.compat.v1.disable_eager_execution()

w=tf.Variable([.2])  # initial weight
b=tf.Variable([-.2]) # initial bias

x= tf.compat.v1.placeholder(tf.float32)  # input
y= tf.compat.v1.placeholder(tf.float32)  # actual output

linear_model=w*x + b    #calculating predicted y

In [22]:
init =tf.compat.v1.global_variables_initializer()  # initalizing variables


s= tf.compat.v1.Session() # setting up session for computation

s.run(init)
print(s.run(w))
print(s.run(b))

[0.2]
[-0.2]


Lets calculate the error after 1st epoch


In [25]:
square_details=tf.square(linear_model-y)
loss=tf.reduce_sum(square_details)
print("loss or error is:")
print(s.run(loss, {x:[1,2,3,4], y:[0,-1,-2,-3]}))

loss or error is:
20.16


error is 20.16 . So lets update the weights and check error again

In [26]:
# now we want to reduce the rmse so we will use gradient descent optimizer

optimizer =  tf.compat.v1.train.GradientDescentOptimizer(0.01) # 0.01 is learning rate
train = optimizer.minimize(loss)
s.run(init)
for i in range(1000):  # 1000 is training epoch
 s.run(train, {x:[1,2,3,4,],y:[0,-1,-2,-3]})

In [27]:
print(s.run([w,b])) # we will get values for weight and bias

[array([-0.99999714], dtype=float32), array([0.99999166], dtype=float32)]


In [28]:
s.run(linear_model,{x:[1,2,3,4],y:[0,-1,-2,-3]})

array([-5.4836273e-06, -1.0000026e+00, -1.9999998e+00, -2.9999969e+00],
      dtype=float32)

In [29]:
s.run(loss,{x:[1,2,3,4],y:[0,-1,-2,-3]})

4.6611603e-11